### PR TITLE
column-mapping: set schemaID/tableID to 0 if numeric suffix is missing

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

(TOOL-1134) Support mixing names with and without numeric suffix.

### What is changed and how it works?

Added an optional *separator* argument to column mapping. The suffix is extracted as *prefix* + *separator* + *suffix*.

If the *name* equals to *prefix* (without the separator), we treat this to mean using a zero numeric suffix, i.e. if `prefix == "money", separator == "_"`:

* `money` → 0
* `money_` → error
* `money_5` → 5
* `money_6` → 6
* `money_7` → 7
* `money_money_money` → error
* `mon` → error
* `not_money` → error

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

 - Need to update the documentation

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@GregoryIan @ericsyh Have we decided to support this?

if so, `money`, `money_` and `mon` are all mapped to `0`, we need to underline it in the doc to avoid misuse.

**Comment by kennytm** [comment]
We could also agree on a list of acceptable separators so that `money` and `money_` works but not `mo`.

**Comment by IANTHEREAL** [comment]
mainly ambiguity problem, I think we can choose one in two way

* define prefix and separator, like prefix is `money` and separator is `_`,  it also means `money`, and `money_` are mapped to `0`
* only define prefix, like prefix is `money`,  `money` is mapped to `0` and ``money_` is invalid

The reason for this is based on the fact that we should provide a definition that is as simple and nature as possible, not easy to make mistakes. how about you? @kennytm @csuzhangxc

**Comment by kennytm** [comment]
The second option seems better. I suggest the following changes:

1. We extend the arguments to 4 elements

    ```yaml
    arguments: [123, 'schemaprefix', 'tableprefix', '_']
    ```

    If the last argument is missing, it defaults to `''`

2. Schema name must be `schemaprefix` or `schemaprefix + sep + schemaID`
3. Table name must be `tableprefix` or `tableprefix + sep + tableID`

**Comment by IANTHEREAL** [comment]
it's fine, and we still allow 3 elements arguments which implies `sep` is `""`

**Comment by kennytm** [comment]
PTAL @csuzhangxc @GregoryIan @ericsyh

File: pkg/column-mapping/column.go [hunk 1]
```diff
@@ -490,22 +495,33 @@ func computePartitionID(schema, table string, rule *Rule) (instanceID int64, sch
 		instanceID = int64(instanceIDUnsign << shiftCnt)
 	}
 
+	var sep string
+	if len(rule.Arguments) > 3 {
```
Comment (csuzhangxc) [inline pkg/column-mapping/column.go]: can we only do this check _once_ for one rule? like do it in `AddRule`/`UpdateRule `, not in `computePartitionID`.

File: pkg/column-mapping/column.go [hunk 2]
```diff
@@ -63,10 +63,12 @@ var Exprs = map[Expr]func(*mappingInfo, []interface{}) ([]interface{}, error){
 	// # 3 table ID (table suffix)
 	// # 4 origin ID (>= 0, <= 17592186044415)
 	//
-	// others: schema = arguments[1] + schema suffix
-	//         table = arguments[2] + table suffix
-	//  example: schema = schema_1 table = t_1  => arguments[1] = "schema_", arguments[2] = "t_"
+	// others: schema = arguments[1]  or  arguments[1] + arguments[3] + schema suffix
```
Comment (IANTHEREAL) [inline pkg/column-mapping/column.go:66]: can you help to update readme? @kennytm

File: pkg/column-mapping/column.go [hunk 3]
```diff
@@ -178,37 +182,33 @@ func (m *Mapping) AddRule(rule *Rule) error {
 	if !m.caseSensitive {
 		rule.ToLower()
 	}
+	if rule.Expression == PartitionID && len(rule.Arguments) == 3 {
```
Comment (IANTHEREAL) [inline pkg/column-mapping/column.go]: is it better to put it into `rule.Ajust`?

**Comment by IANTHEREAL** [comment]
Rest LGTM

**Comment by IANTHEREAL** [comment]
LGTM @csuzhangxc PTAL

**Review by csuzhangxc — APPROVED** [review]
LGTM. Could you help to update the docs-cn or docs?

**Comment by kennytm** [comment]
pingcap/docs#1166.

**Pull request merged**